File: python/dsa/algorithms/matrix/rotate-image.py
def rotate(mat):
    n = len(mat)
    # 1. Transpose
    for i in range(n):
        for j in range(i, n):
            mat[i][j], mat[j][i] = mat[j][i], mat[i][j]
    # 2. Reverse rows
    for i in range(n):
        left, right = 0, n - 1
        while left < right:
            mat[i][left], mat[i][right] = mat[i][right], mat[i][left]
            left, right = left + 1, right - 1
    return mat

File: go/dsa/algorithms/matrix/rotate-image.go
func rotate(mat [][]int) {
    n := len(mat)
    // 1. Transpose
    for i := 0; i < n; i++ {
        for j := i; j < n; j++ {
            mat[i][j], mat[j][i] = mat[j][i], mat[i][j]
        }
    }
    // 2. Reverse rows
    for i := 0; i < n; i++ {
        left, right := 0, n-1
        for left < right {
            mat[i][left], mat[i][right] = mat[i][right], mat[i][left]
            left++; right--
        }
    }
}